### ISSUE TITLE
RFC: special: Return type of special functions that return derivatives.

### ISSUE THREAD

**Issue by steppi** [scipy/scipy#21782]
In relatively recent work (https://github.com/scipy/scipy/pull/20539), @izaid developed a more consistent API for the Legendre functions, and developed tools for evaluating the recurrences used in their computation that simplified the implementation and reduced code duplication. These functions are able to return derivatives of Legendre functions up to order 2, and this is done through a keyword argument, `diff_n`, which takes values `0`, `1`, or `2`. The output of the function is then a tuple of length 1, 2, or 3 with the following values:

|diff_n| result |
|--------|----------|
| 0       |      (function value, )      |
| 1       |      (function value, gradient)      |
| 2       |      (function value, gradient, hessian)    |

This was discussed in the discourse post, https://discuss.scientific-python.org/t/naming-convention-for-generalized-ufuncs-in-special/1144, and in the issue, https://github.com/scipy/scipy/issues/20448. None of this has made it into a release yet. The functions with this behavior are all new functions, and we plan to deprecate the older Legendre function API.

Having a keyword argument control the number of output values is not ideal. In a comment on the discourse post mentioned above, @rgommers said:

*Some things are worse than others, obviously. Here is how I’d order them from most to least impactful (in a negative way) for behavior based on the value of a keyword:*

* Returning a different number of output values
* Returning an array with a different dtype
* Returning an array with a different number of dimensions

At the time #20539 went in, it would have been difficult to take another path, but since then, @izaid has done additional work improving how derivatives are computed for special functions, adding forward autodifferentiation in https://github.com/scipy/scipy/pull/21483. This makes computation of derivatives much more flexible (previously it was done in Zhang and Jin's code by manually differentiating code line by line). With this change, @izaid has informed me that it would be much easier to have legendre functions, and other special functions that return derivatives, each return a single array.

For functions of two variables $f(x, y)$, the idea is to return, for scalar input a 2d array `out` such that

$$\mathrm{out}\left[i, j\right] = \frac{\partial^{i+j}f}{\partial x^{i} \partial y^{j}}$$

and for array input to add two new dimensions for the derivatives. For functions of a single variable, it's straightforward what to do, and for functions with more than two variables, I hope it's clear how to generalize this.

There's a question of whether dimensions for the derivatives should be appended to the beginning or end of the output. I think having them at the front would be more consistent with what has already been done with the legendre functions, but will let @izaid weigh in on this.

We branch in around 5 weeks, so there's still time to change this before the new Legendre functions make it into the public API. I know I would prefer to see them return a single array (with variable shape) rather than a variable number of outputs depending on a kwarg.

**Comment by steppi** [comment]
Thanks @izaid. I agree now about appending the new dimensions to the end.

**Comment by steppi** [comment]
> I would encourage us to adjust our recent changes to that, by changing the keyword from `diff_n = n` to something like `diff_orders = (n0, n1, ...)` but with `diff_orders = n` as a shortcut for `(n,)`.

@izaid, can you elaborate on this? I'm not entirely sure how `diff_orders` would behave.

**Comment by mdhaber** [comment]
Assuming scalar inputs, do I understand correctly that a) the maximum of `diff_orders` would control the length of these two new dimensions and b) each element in an array of `diff_orders` would add numbers in an anti-diagonal? 

What about elements not specified by `diff_orders`, such as the whole lower right triangular (which I assume would have nothing in it, even with `diff_n`)?

The layout of Hessians as a matrix is useful, but IIUC this would not do that 

Are you considering higher higher dimensions? Without seeing this proposal, I would have assumed that if `diff_n=3` we'd get three extra dimensions, each with length equal to the number of input variables.

**Comment by izaid** [comment]
> Assuming scalar arguments, do I understand correctly that a) the maximum of `diff_orders` would control the length of these two new dimensions and b) each element in an array of `diff_orders` would add numbers in an anti-diagonal?

Yes, so what is proposed here is as follows.

- For a scalar function of one argument `x`, `out[..., i]` is the `i`th derivative.
- For a scalar function of two arguments `x` and `y`, `out[..., i, j]` is the `i`th derivative of `x` and the `j`th derivative of `y`.
- For a scalar function of three arguments `x`, `y`, and `z`, `out[..., i, j, k]` is the `i`th derivative of `x`, `j`th derivative of `y`, and `k`th derivative of `z`.
- And so on.

Then `diff_orders` specifies the shape of those derivative dimensions.

> What about elements not specified by `diff_orders`, such as the whole lower right triangular (which I assume would have nothing meaningful in it, even with `diff_n`)?

This isn't the Hessian, so there's no duplication of values and all elements are meaningful.

> The layout of Hessians as a matrix is useful (e.g. is it positive definite?) but IIUC this would not do the usual thing (at least in the context of optimization) for Hessians?

Indeed! Again, our first stab at this actually did do the usual thing for Hessians. That turned out to be a very awkward idea for a ton of reasons, ranging from having multiple outputs to passing around tuples in C++ code. This has emerged as a much more natural solution.

Of course, you can convert this to the Hessian just by indexing `out[..., i, j]` where `i = [[2, 1], [1, 0]]` and `j = [[0, 1], [1, 2]]`.

> Are you considering higher derivatives than two? Without seeing this proposal, I would have assumed that if `diff_n=3` we'd get three extra dimensions, each with length equal to the number of input variables. But I don't know how higher derivatives are handled in other software; I've only worked up to two.

Yes, as per the above. We probably won't implement that at the moment, but we want to leave it open for the future.

**Comment by mdhaber** [comment]
> That turned out to be a very awkward idea for a ton of reasons... Of course, you can convert this to the Hessian just ...

So it's very easy for the user to do, but very awkward to do in C++?

> Yes, so what is proposed here is as follows.

I think I understood that. My questions were about the implications of that scheme. To be even more concrete?

Suppose `diff_orders=[1, 3]` is passed to a function with two input arguments.
1. What is `out.shape[-2:]`?
2. What is located at `out[..., 2, 0]`?
3. What is located at `out[..., 3, 3]`, if it exists?

Answers, from what I can tell:
-  I assume the shape of the last two dimensions is `(4, 4)` so we can index, say, $i = 0, j =3$ to get the third partial derivative with respect to the second variable and all partials such that $i + j = 3$.
- That would be $\frac{d^2f}{dx^2}$, but we didn't request `2` in `diff_orders`, so it's unclear whether that partial derivative will be calculated or not.
- That would be $\frac{d^6}{dx^3 dy^3}$, but again, we didn't request `6` in `diff_orders`, so I would assume I would not get derivatives of sixth total order.

**Comment by izaid** [comment]
Thanks for the questions! Thoughts are as follows, assuming derivative dimensions are at the end.

> Suppose `diff_orders=[1, 3]` is passed to a function with two input arguments.
> 
>     1. What is `out.shape[-2:]`?

It is `(1, 3)`.

>     2. What is located at `out[..., 2, 0]`?

That's out of bounds, but `out[..., 0, 2]` would the second derivative with respect to the second argument.

>     3. What is located at `out[..., 3, 3]`, if it exists?

It doesn't exist.

> Answers, from what I can tell:
> 
>     1. I assume the shape of the last two dimensions is `(4, 4)` so we can index, say, 
>          i
>          =
>          0
>          ,
>          j
>          =
>          3
>         to get the third partial derivative with respect to the second variable and all partials such that 
>          i
>          +
>          j
>          =
>          3
>         (along the anti-diagonal).
> 
>     2. That is supposed to be 
>          
>            
>              
>                d
>                2
>              
>              f
>            
>            
>              d
>              
>                x
>                2
>              
>            
>          
>        , but we didn't request `2` in `diff_orders`, so it's unclear whether that partial derivative will be calculated or not.
> 
>     3. That is supposed to be 
>          
>            
>              
>                d
>                6
>              
>              f
>            
>            
>              d
>              
>                x
>                3
>              
>              d
>              
>                y
>                3
>              
>            
>          
>        , but again, we didn't request `6` in `diff_orders`, so I would assume I would not get derivatives of sixth total order.
> 
> 
> If we only want to think about 2nd order for now, LMK and I can change the questions accordingly.

I think your answers are correct.

**Comment by mdhaber** [comment]
> It is (2, 4)

OK, that makes things more clear. When you wrote "by changing the keyword from `diff_n = n` to something like `diff_orders = (n0, n1, ...)`", I understood `n0`, `n1`, etc., to represent a list of *total* derivative orders, not a sequence of maximum orders w.r.t. to each variable.

Now that I understand the format, I would just ask whether that is a good format for the user.

I'm familiar with wanting a gradient. To get that, I'd pass `diff_orders=(1, 1)`, and extract the main anti-diagonal. $\frac{d^2f}{dx dy}$ would also be calculated, but I wouldn't use it unless I happened to need that second-order derivative but not the others.

I'm also familiar with wanting a Hessian. To get that, I'd pass `diff_orders=(2, 2)` and extract elements from the main anti-diagonal into a matrix. Several higher-order derivatives would be calculated that I may not need.

If it's more common to want all partials up to a certain maximum order for each variable, this sounds like a good format. If it's more common to want gradients and Hessians, we might want something else.

**Comment by izaid** [comment]
Yep, that's basically it! All I would add is that underneath, what we are describing is the way the calculation is now done and how is laid out in memory. The fact that we are still returning func, jac, and hessian for (say) `legendre_p` is because I've put in the aforementioned indexing to maintain it. To me, that's another reason to adjust things, we shouldn't be doing extra work unless the user wants it explicitly.

**Comment by steppi** [comment]
Ah, I get it now. Since we’re using dual numbers for the derivatives, it’s natural that we end up computing them up to some fixed order in each variable, so even when we return the just the function, gradient, and hessian we still end up computing some higher thinks internally.

**Comment by mdhaber** [comment]
> because I've put in the aforementioned indexing to maintain it. To me, that's another reason to adjust things, we shouldn't be doing extra work unless the user wants it explicitly.

OK, so the proposal is to *undo* the aforementioned indexing in `legendre_p` (and not add it going forward) and return exactly what autodiff is calculating naturally.

Crazy idea: return the array as you propos - but extend it with methods (or properties) that do the indexing when requested. And while we're at it, add a property for the internal accuracy estimate, if there is one. (I want that for other reasons.)

**Comment by izaid** [comment]
> > 
> 
> OK, so the proposal is to _undo_ the aforementioned indexing in `legendre_p` (and not add it going forward) and return exactly what autodiff is calculating naturally.

Exactly! If desired, we can add a helper function to go between the formats as well, maybe something like that could live in `differentiate`.

**Comment by steppi** [comment]
Just a thought. If I understand correctly, right now we define $n$ variable duals as duals with coefficients that are $n - 1$ variable duals. This makes them easy to define, but then you have this thing we're seeing where you ask for order 2 and get $$\frac{\partial ^4 f}{dx^2dy^2}$$

using the notation $$x + dx + dx^2$$, $$y + dy + dy^2$$ for duals, if you want order 2, I think it's usually the case that $$dx^{i}dy^{j} = 0$$ when $$i + j > 2$$. That's at least how things work in nonstandard analysis and non-rigorous Newtonian calculus.

So is it actually natural that, for example, you can get higher order derivatives than just the hessian when calculating with duals for order 2, or is it just an implementation detail?

**Comment by izaid** [comment]
> Just a thought. If I understand correctly, right now we define n variable duals as duals with coefficients that are n − 1 variable duals. This makes them easy to define, but then you have this thing we're seeing where you ask for order 2 and get ∂ 4 f d x 2 d y 2 as something nonzero.
> 
> using the notation x + d x + d x 2 , y + d y + d y 2 for duals, if you want order 2, I think it's usually the case that d x i d y j = 0 when i + j > 2 . That's at least how things work in nonstandard analysis and non-rigorous Newtonian calculus.
> 
> So is it actually natural that, for example, you can get higher order derivatives than just the Hessian when calculating with duals for order 2, or is it just an implementation detail?
> 
> This can make a big difference in the amount of storage needed per dual number too. If you have a 4 variable function like `hyp2f1` and ask for derivatives up to order 2 in each variable, then taking up to order 2 in each variable gives you 3 × 3 × 3 × 3 = 81 entries per dual, but taking up to total order two, you just need ( 6 4 ) = 10 entries per dual.

This is all correct. Yes, there's a difference in memory, for functions with many arguments it can be more. We generally won't have special functions with many arguments. It's just a question of trade-offs, and we don't really have another option.

**Comment by steppi** [comment]
> This is all correct. Yes, there's a difference in memory, for functions with many arguments it can be sizable. We generally won't have special functions with that many arguments, the vast majority have 1 argument, a small number have 2, and a very small number have more than 2. It's just a question of trade-offs, and we don't really have another option.

Yeah, I agree it would be challenging to implement duals with the usual convention that everything with total order greater than $$n$ is zeroed out for arbitrary orders and numbers of variables [^1], and don't think we should pursue this now. I'm thinking more in API terms. What I'm thinking is that right now it's an implementation detail that in the C++ we specify what is essentially `diff_orders=(n1, n2, ...)` and get an  $$n_1 \times n_2 \times \cdots $$ array output, but in math terms, it isn't really the most natural thing to do to take partial derivatives up to a certain order in each component. I'm starting to wonder if it's not the best move to tie the output format to an implementation detail that is a suboptimal but acceptable compromise. 

[^1] Though we could hardcode things just for the orders and numbers of variables we need.

**Comment by izaid** [comment]
> > This is all correct. Yes, there's a difference in memory, for functions with many arguments it can be sizable. We generally won't have special functions with that many arguments, the vast majority have 1 argument, a small number have 2, and a very small number have more than 2. It's just a question of trade-offs, and we don't really have another option.
> 
> Yeah, I agree it would be challenging to implement duals with the usual convention that everything with total order greater than n is zeroed out for arbitrary orders and numbers of variables [1](#user-content-fn-1-695e48c35c9ecc69ef94686252a7ae2e), and don't think we should pursue this now. I'm thinking more in API terms. What I'm thinking is that right now it's an implementation detail that in the C++ we specify what is essentially `diff_orders=(n1, n2, ...)` and get an n 1 × n 2 × ⋯ array output, but in math terms, it isn't really the most natural thing to do to take partial derivatives up to a certain order in each component. I'm starting to wonder if it's not the best move to tie the output format to an implementation detail that is a suboptimal but acceptable compromise.
> ## Footnotes
> 
>     1. Though we could hardcode things just for the orders and numbers of variables we need. [↩](#user-content-fnref-1-695e48c35c9ecc69ef94686252a7ae2e)

Yes, so from a Python perspective I'm happy to adopt the simplest API possible. And yes, we could think of the duals as an implementation detail.

The only thing is since we are working with arrays, it's not like we can just drop some values. Either we do the  `n1 × n2 × ⋯` output or we do the `func, jac, hessian` output, but that is merely indexing as extra overhead.

**Comment by mdhaber** [comment]
How common is that you would want the higher derivatives that the proposed scheme offers without the user also needing the other derivatives of the same total order? 

Also, do you have a sense of how often you would access the partials individually vs how often you would want the gradient, hessian

**Comment by steppi** [comment]
@izaid, right it comes down to either `n1 x n2 x ...` or `func, jac, hessian`. I wasn't paying sufficient attention before and have only now realized this isn't a slam dunk in favor of one or the other, but something that needs to be thought out. 

> And yes, we could think of the duals as an implementation detail.

I think the implementation detail in question here is that what we have isn't what would typically be considered multivariate duals of $k$ variables of order $n$, but a different algebraic structure which is easier to set up.

I'm on the fence on this one still, and want to make sure both options are carefully considered.

**Comment by steppi** [comment]
Having thought about it. I think my preference is to stick with `func`, `jac`, `hessian` on the Python side. Like @mdhaber, I'm struggling to think of reasons why someone might want only some of the higher order partials. I think `func`, `jac`, `hessian` will be simpler for most users to work with, both mechanically and conceptually, and again I think it's just an implementation detail that internally we compute partials up to order $n_i$ in variable $x_i$ for each $i$, so I don't think it's necessarily important or desirable to match what's going on internally in the output.

**Comment by izaid** [comment]
> Having thought about it. I think my preference is to stick with `func`, `jac`, `hessian` on the Python side. Like @mdhaber, I'm struggling to think of reasons why someone might want only some of the higher order partials. I think `func`, `jac`, `hessian` will be simpler for most users to work with, both mechanically and conceptually, and again I think it's just an implementation detail that internally we compute partials up to order n i in variable x i for each i , so I don't think it's necessarily important or desirable to match what's going on internally in the output.

I think I'm also nearly ready to agree with this, but want to think a bit more.

**Comment by fancidev** [comment]
It seems to me `val, jac, hess = f(...)` feels the most "natural".

How would you like to address the potentially  non-ideal situation of having a function with different return type depending on the parameter value?

**Comment by mdhaber** [comment]
> How would you like to address the potentially non-ideal situation of having a function with different return type depending on the parameter value?

I'm less uncomfortable with returning a different length sequence of outputs as long as the length and the elements themselves are controlled by an input array (rather than a scalar). *In a sense*, that's not so unusual.
```python3
import numpy as np
from scipy import stats
u1, u2, u4 = stats.moment(np.arange(10), order=[1, 2, 4])
```

If a stateful object that can use previously calculated derivatives to calculate higher derivatives is out of the question, I think I'd prefer this, at least over a single integer that specifies the number of derivatives.

**Comment by steppi** [comment]
> I'm less uncomfortable with returning a different length sequence of outputs as long as the length and the elements themselves are controlled by an input array (rather than a scalar). _In a sense_, that's not so unusual.

I'm not sure it's worth the trouble. Since `func`, `jac`, and `hess` all have different shapes in the multivariate case and we can't just tile them up into an array, would we require `order` to be 1d and return a tuple with the same length as `order`? Since we typically can't compute `hess` without computing `jac` and can't compute `jac` without computing `func`, I don't think we really gain anything by being able to do stuff like

```python
func, hess = sph_harm_y(n, m, theta, phi, order=[0, 2])
```

vs

```python
func, _, hess = sph_harm_y(n, m, theta, phi, diff_n=2)
```

and I'm also not sure what we gain by being able to return duplicate copies of the return arrays, or being able to return them in a different order.

@izaid, do you have any thoughts?

**Comment by mdhaber** [comment]
> don't think we really gain anything

I agree we don't gain anything. It's just a matter of the least potentially surprising interface assuming that users are used to the *shape* - not the value - of the input governing the shape of the output.

**Comment by steppi** [comment]
OK, yeah, I think having array orders is a little unusual too though. It doesn't broadcast with the other inputs, it controls the length of a container of arrays, with the individual arrays potentially having different shapes. But what you say makes sense though, and shouldn't be that hard to implement, and the work would be all on the Python side, so wouldn't end to hold up the `xsf` migration I'm working on right now. 

If someone passes an array with one unsupported order, I guess we just error out completely, even if we could compute the other derivatives? 

@izaid, do you think it's fine if we do what @mdhaber proposes?

**Comment by mdhaber** [comment]
> It doesn't broadcast with the other inputs

Yeah, I was going to mention that as a downside. And it is a real downside in `percentile`/`quantile`, and something I'd like to fix in similar SciPy functions; those really should follow normal broadcasting rules. Here, it's not really possible to follow normal broadcasting rules without returning an object array; that's why I didn't suggest it (but I did consider it).

> I guess we just error out completely, even if we could compute the other derivatives?

I think NaNs are better. Invalid input -> NaN output.

**Comment by steppi** [comment]
> I think NaNs are better. Invalid input -> NaN output. You can warn or not as long as it's documented, I think. At least that's the opinion in `stats`.

OK, yeah, I was concerned about the ambiguity between something returning NaNs because the derivative genuinely doesn't exist at a point, or NaNs because we don't offer derivatives of a given order. We already ahve this same problem in `special` for some functions though, where we return NaN in regions where we just don't have an implementation, `wright_bessel` for instance. If it's documented then I guess it should be fine.

**Comment by fancidev** [comment]
I suppose one challenge with parameter value affecting the return type is for static typing, but that might be resolvable if the parameter is passed as a literal. So the signature could be

```python 

@typing.overload
def f(args) -> array: ...

@typing.overload
def f(args, diff_n: Literal[0]) -> array: ...

@typing.overload
def f(args, diff_n: Literal[1]) -> tuple[array, array]: ...

@typing.overload
def f(args, diff_n: Literal[2]) -> tuple[array, array, array]: ...
```

I'm not entirely sure whether static type checkers support such overloads. Maybe typing expert @jorenham could share some ideas?

**Comment by mdhaber** [comment]
> We already have this same problem in special

Yeah, it's all over the place. It always baffles me that there are so many representations, but we can't easily treat them like different sentinel values in Python.

NaNs, NaNs, every where.
Quadrillion floats, I think?
NaNs, NaNs, every where,
Nor any are distinct.

**Comment by izaid** [comment]
@mdhaber @steppi Checking in here. As the person who has written all the derivative code, and knows how much work it is, I don't think we should be doing changing `diff_n` to accept scalars.

It's been a monumental amount of work even to get it to this point. The reason why I was interested in returning a single array instead of a `func, jac, hess` tuple, as originally discussed here, was that it would simplify things greatly. The argument has now been made that `func, jac, hess` is more common and that the duals are an implementation detail. I actually accept that, but it means more work to set this up properly and maintain it. We're also struggling with maintainer time at the moment and, so far, the only person who has implemented the derivatives is me (and my time is really tight now as @steppi knows).

For all those reasons, I think we should just leave `diff_n` as is and I'm against making it something other than a scalar. If there is serious demand from others, we can revisit in the future. I think this is a good compromise, as I'm still not 100% sure we shouldn't just return the dual layout exactly as it is.

**Comment by mdhaber** [comment]
> As the person who has written all the derivative code, and knows how much work it is,

I just had in in mind applying a decorator at the Python level - not any changes to your code.

```python3
import numpy as np
def decorator(f):
    def wrapper(*args, n=0, **kwargs):
        single_n = np.isscalar(n) 
        ns = [n] if single_n else n
        diff_n = min(max(ns), 2)  # if 2 is the max order
        values = f(*args, diff_n=diff_n, **kwargs)
        out = tuple(values[i] if (0 <= i <= diff_n) else np.nan for i in ns)
        return out[0] if single_n else out
    return wrapper

@decorator
def f(diff_n):  
    # pretend this returns derivatives of orders up to/including `diff_n`
    return np.arange(diff_n + 1)

f(n=1)  # 1
f(n=-1)  # nan
f(n=[0, 2, 5])  # (0, 2, nan)
```

But that's my last two cents here unless more is requested.

**Comment by izaid** [comment]
@mdhaber Thanks again for the thoughts! It is appreciated and good to consider what's possible.

I still think we should just leave things as is. I actually would like to minimise the Python code for these functions and, if possible, turn as much as possible into vanilla ufuncs. That was another reason why I thought returning the duals themselves would be valuable, as it meant a more natural fit into ufuncs.

I do agree that, when working with derivatives, what seems naturally wanted is `func, jac, hess`. So for that reason I think we should everything as is with `diff_n` and, in time, I'll try to push this more into the C++ code.

**Comment by jorenham** [comment]
> I suppose one challenge with parameter value affecting the return type is for static typing, but that might be resolvable if the parameter is passed as a literal. So the signature could be
> 
> ```python
> @typing.overload
> def f(args) -> array: ...
> 
> @typing.overload
> def f(args, diff_n: Literal[0]) -> array: ...
> 
> @typing.overload
> def f(args, diff_n: Literal[1]) -> tuple[array, array]: ...
> 
> @typing.overload
> def f(args, diff_n: Literal[2]) -> tuple[array, array, array]: ...
> ```
> 
> I'm not entirely sure whether static type checkers support such overloads. Maybe typing expert @jorenham could share some ideas?

This example is a valid one, and there's indeed no better way to annotate "a tuple of length `n`" at the moment.

It's always a good thing to keep static typing in mind when it comes to designing a new API. But there's no need to worry about the details too much; that's what [`scipy-stubs`](https://github.com/jorenham/scipy-stubs) is for 😉.

**Comment by steppi** [comment]
> The argument has now been made that `func, jac, hess` is more common and that the duals are an implementation detail

@izaid. I want to clarify that what I consider an implementation detail is that the dual-like numbers used cause the entire matrix to be filled, because products of distinct variables with total order $> n$ don't vanish, only those with a maximum order $> n$.

> Believe it or not, I don't have a strong opinion that this should accept an array - I just believe it's more consistent with other functions and would satisfy the concerns (expressed by others, IIRC) against input value controlling output shape.

I don't really have a strong opinion about it either. Taking a 1d array for `diff_orders` does seem more consistent with the behavior of other functions in SciPy, but I guess it's a matter of whether it's worth the effort at this point. I'm not sure it's that important to minimize Python code or harmonize with the underlying C++ library when it comes to making Python level API decisions. There's not really a shortage of bandwidth for maintaining pure Python code in SciPy, and I think user experience should come first so long as the maintenance burden incurred is not excessive.

We branch on December 6th according to this [plan](https://discuss.scientific-python.org/t/proposed-release-schedule-for-scipy-1-15-0/1502). I'm trying to think if it's even practical to make the change before then. I spent a little time looking into what it would take.

 I think changing to a kwarg `diff_orders` taking array inputs would best be done at the Python level. and it wouldn't even need a decorator, everything happens in the `MultiUFunc` class, and there would only be a few changes needed. Here's where the output gets created currently:

https://github.com/scipy/scipy/blob/862fce8e226962138597a916944bb13b67aadd16/scipy/special/_multiufuncs.py#L135-L139


and some logic like in @mdhaber's decorator could appear here. Prior to this, everywhere that `diff_n` is currently being used could be replaced with`max(diff_orders)`. It doesn't seem too bad; updating the documentation seems like it could be the most painful part. I think the added overhead would be negligible. There'd still be some questions:

1.  What to do if someone passes in an unsupported order: (put an array of NaNs of the correct output shape in the proper spot in the tuple. even if it ends up being huge? It might be better just to raise). 
2. If redundant orders are given in `diff_orders` does it make more sense to return views into the same array (like @mdhaber's decorator), or copies. The former is more efficient but maybe could be confusing if someone wants to modify the outputs somehow.

@izaid, I would see this as a Python only change that shouldn't need any work from you, and wouldn't require you to change anything you have planned at the C++ level. If say, @mdhaber submitted a PR to make this change, and the incurred performance overhead is negligible, would you object? I suspect the incurred maintenance burden would be small. Then again, if this is an improvement, it's only a small one. If you just really don't want this to happen, and are not just worried about having more work heaped on you, I'll respect that though, and we can stick with the current status quo.

As an aside, I don't know much about static typing in Python yet, but the examples for typing such things are very interesting. 

[^1]: I don't think there is such a return type though.

**Comment by izaid** [comment]
If we are going to go this, I'd rather we go back to my original proposal of just returning the exact array as calculated by the duals. I think we are overcomplicating things by a hell of a lot and I don't see a strong argument for doing this. So, yes, I (strongly) object - sorry guys!

**Comment by fancidev** [comment]
1. I suspect it's currently impossible to type annotate meaningfully the signature if passing `diff_orders` as suggested. For example, with

```Python
val, jac, hess = f(x, y, diff_n=2)
```

it's possible to type annotate in a way such that the type checker / IDE is able to deduce the type and *shape* of `val`, `jac`, and `hess` given the shape of `x` and `y`. However, it cannot if passing `diff_orders=[0,1,2]`.

The "logic" of type annotation is somewhat similar to C++'s templates. `diff_n` can be thought of as a template argument of type `int`, and the function's return type may vary depending on its value. For this to work, `diff_n` must be passed a literal, much like a C++ template must be instantiated with constexpr (non-type) argument.

If this is to work with `diff_orders`, first `diff_orders` must be passed a tuple literal, and second the type annotation machinery must support "variadic" tuple literals, which I suspect is not currently the case.

Besides, the use case here is different from `percentile` and `moment` in what I think to be an important aspect: For those functions, the parameter is essentially "vectorizing" the output. But here, different order corresponds to return value of different shape. I feel `diff_n` appears to suggest weaker analogy  to vectorization and might be preferrable.

Another point seems not discussed yet: What if the function is univariate? Do you want to return val, jac, hess as a tuple? Or do you want to combine them into one array as now all derivatives are scalar so there is no obstacle for shape "squeezing" or type annotation?

**Comment by jorenham** [comment]
> If this is to work with `diff_orders`, first `diff_orders` must be passed a tuple literal, and second the type annotation machinery must support "variadic" tuple literals, which I suspect is not currently the case.

Variadic type parameters are a thing, but currently there's indeed no way to map e.g. `tuple[int, ...]` to `tuple[array, ...]` while maintaining its size. Something like that would require [variadic higher-kinded type parameters](https://github.com/python/typing/issues/548).

**Comment by steppi** [comment]
> If we are going to do this, I'd rather we go back to my original proposal of just returning the exact array as calculated by the duals. I think we are overcomplicating things by a hell of a lot and I don't see a strong argument for doing this. So, yes, I (strongly) object - sorry guys!

No worries. I accept that. Let's keep it as is.

Thanks for the thoughts @fancidev. I agree that the scalar `diff_n` actually working with type annotations is a good argument in its favor. I think @izaid and I did talk about having these functions output a single array in the univariate case and only returning a tuple for the multivariate case at some point. I guess it's a matter of whether or not consistency is more important.
